spec/unit/indirector/catalog/compiler_spec.rb

```diff
@@ -414,6 +414,30 @@ def a_request_that_contains(facts)
     end
   end
 
+  describe "after finding nodes, in regard to the server's Puppet Enterprise version" do
+    pe_version_file = '/opt/puppetlabs/server/pe_version'
+
+    it "should not add 'pe_version' when FOSS" do
+      allow(File).to receive(:readable?).with(pe_version_file).and_return(false)
+      @request = Puppet::Indirector::Request.new(:catalog, :find, node_name, nil)
+      allow(compiler).to receive(:compile)
+      allow(Puppet::Node.indirection).to receive(:find).with(node_name, anything).and_return(node)
+      expect(node).to receive(:merge).with(hash_excluding('pe_version'))
+      compiler.find(@request)
```

Comment:
Relying on `:merge` is a little fragile. I'd recommend not stubbing out as much, compile a catalog, and compare what it produced:

```ruby
allow(File).to receive(:read).with(pe_version_file).and_return('1.2.3')
catalog = compile_to_catalog('notify { "$pe_version": }', node)
expect(catalog.resource('notify', '1.2.3')).to be_nil
```

And the case below is just `expect(catalog.resource('notify', '1.2.3')).to be`

Reply:
I modeled testing `pe_version` after the `describe "after finding nodes" do` block on L391.

This passes:

```
    include PuppetSpec::Compiler
    let (:pe_version_file) { '/opt/puppetlabs/server/pe_version' }

    it "should not add 'pe_version' when FOSS" do
      catalog = compile_to_catalog('notify { "PE Version: ${pe_version}": }', node)
      expect(catalog.resource('notify', 'PE Version: ')).to be
    end
```

But this fails:

```
    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')
      catalog = compile_to_catalog('notify { "PE Version: ${pe_version}": }', node)
      expect(catalog.resource('notify', 'PE Version: 2019.1.2')).to be
    end
```

With:

```
    Failure/Error: expect(catalog.resource('notify', 'PE Version: 2019.1.2')).to be
       expected nil to evaluate to true
```

This passes:

```
    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')
      catalog = compile_to_catalog('notify { "PE Version: 2019.1.2": }', node)
      expect(catalog.resource('notify', 'PE Version: 2019.1.2')).to be
    end
```

Suggesting the `allow` statements are not working (they do work, and each is necessary, in the original commit). 

Any advice? :)

Reply:
Sorry, yeah, you'll need the the allow `readable?` and `executable?` statements.

Reply:
I do allow [readable?, zero?, and read] (don't need executable?) ...

```
    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')
      catalog = compile_to_catalog('notify { "PE Version: ${pe_version}": }', node)
      expect(catalog.resource('notify', 'PE Version: 2019.1.2')).to be
    end
```

 But File is not receiving them:

```
    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')
      catalog = compile_to_catalog('notify { "PE Version: ${pe_version}": }', node)
expect(File).to have_received(:readable?).with(pe_version_file)
      expect(catalog.resource('notify', 'PE Version: 2019.1.2')).to be
    end
```

```
[tom.kishel@tkishel puppet]$ bundle exec rake spec TEST=spec/unit/indirector/catalog/compiler_spec.rb:417

...

     Failure/Error: expect(File).to have_received(:readable?).with(pe_version_file)
     
       (File (class)).readable?("/opt/puppetlabs/server/pe_version")
           expected: 1 time with arguments: ("/opt/puppetlabs/server/pe_version")
           received: 0 times
     # ./spec/unit/indirector/catalog/compiler_spec.rb:432:in `block (3 levels) in <top (required)>'
```

Whereas the original version, File is receiving them:

```
    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')
      @request = Puppet::Indirector::Request.new(:catalog, :find, node_name, nil)
      allow(compiler).to receive(:compile)
      allow(Puppet::Node.indirection).to receive(:find).with(node_name, anything).and_return(node)
      expect(node).to receive(:merge).with(hash_including('pe_version' => '2019.1.2'))
      compiler.find(@request)
expect(File).to have_received(:readable?).with(pe_version_file)
    end
```

```
[tom.kishel@tkishel puppet]$ bundle exec rake spec TEST=spec/unit/indirector/catalog/compiler_spec.rb:417

Finished in 0.15214 seconds (files took 0.88983 seconds to load)
2 examples, 0 failures
```

The stubs are snubbed!

Reply:
ok, my bad. I forgot that `compile_to_catalog` just calls `Puppet::Parser::Compiler`, but doesn't perform any of the post-processing that the compiler terminus does. So I'd recommend the following instead:

```ruby
  describe "after finding nodes, in regard to the server's Puppet Enterprise version" do
    include PuppetSpec::Compiler

    let(:pe_version_file) { '/opt/puppetlabs/server/pe_version' }
    let(:request) { Puppet::Indirector::Request.new(:catalog, :find, node_name, nil) }

    before :each do
      Puppet[:code] = 'notify { "version: $pe_version": }'
    end

    it "should not add 'pe_version' when FOSS" do
      allow(Puppet::Node.indirection).to receive(:find).with(node_name, anything).and_return(node)
      catalog = compiler.find(request)

      expect(catalog.resource('notify', 'version: 2019.1.2')).to be_nil
    end

    it "should add 'pe_version' when PE" do
      allow(File).to receive(:readable?).with(pe_version_file).and_return(true)
      allow(File).to receive(:zero?).with(pe_version_file).and_return(false)
      allow(File).to receive(:read).with(pe_version_file).and_return('2019.1.2')

      allow(Puppet::Node.indirection).to receive(:find).with(node_name, anything).and_return(node)
      catalog = compiler.find(request)

      expect(catalog.resource('notify', 'version: 2019.1.2')).to be
    end
  end
```

Reply:
Thanks, I appreciate your patience and time!